MAINT: Translate x86-qsort.dispatch.c.src to C++ using templates.

serge-sans-paille opened the pull request:
Using a approach close to engine like xsimd here, to abstract the vectorized
code.

seberg commented:
Have not looked at it, but could you make sure to rename/`git mv` the file rather than delete and create a new one again?

serge-sans-paille committed:
road-to-cxx: x86-qsort.dispatch.c.src -> x86-qsort.dispatch.cpp

Actual c++ification.

Also fix a slight bug in numpy/distutils/ccompiler_opt.py when intermediate
directory is missing.

serge-sans-paille committed:
road-to-cxx - x86-qsort.dispatch.c.src -> x86-qsort.dispatch.cpp

This is an intermediate step without changes so that git can keep track of the
move.

serge-sans-paille commented:
> Have not looked at it, but could you make sure to rename/`git mv` the file rather than delete and create a new one again?

I Moved it, then modified it to do the conversion. Split that in two steps.

seberg commented:
Ah, thanks.  Is github just weird about showing moved files only well if the diffs are small/moved without change?

serge-sans-paille committed:
Make common/simd/avx512/memory.h compatible with C++

serge-sans-paille commented:
green and ready for review

serge-sans-paille commented:
ping :-) This one was a bit more difficult to handle than the others, but thanks to the (thin) simd abstraction, everything went well.

mattip commented on numpy/distutils/ccompiler_opt.py:
```diff
@@ -2551,6 +2551,8 @@ def _generate_config(self, output_dir, dispatch_src, targets, has_baseline=False
             pass
 
+        os.makedirs(os.path.dirname(config_path), exist_ok=True)
+
```
Any idea why this is needed now?

serge-sans-paille replied:
yeah, it happens when there's no `.c.src` remaining in a setup.py subdirectory, but the system somehow needs that subdirectory to be created.

mattip commented on numpy/core/src/common/simd/avx512/reorder.h:
```diff
@@ -214,13 +214,13 @@ NPY_FINLINE npyv_u16 npyv_rev64_u16(npyv_u16 a)
 
 NPY_FINLINE npyv_u32 npyv_rev64_u32(npyv_u32 a)
```
Was this (and `npyv_rev64_f32`) ever called?

serge-sans-paille replied:
I don't know, I'm just a translator ;-)

r-devulap commented:
> ping :-) This one was a bit more difficult to handle than the others, but thanks to the (thin) simd abstraction, everything went well.

The patch looks great! quick question w.r.t the SIMD abstraction: does it in anyway overlap with abstracting universal intrinsics using C++ templates? I am not sure what the scope of that work is, so it might not be a relevant question at all :)

serge-sans-paille commented:
> > ping :-) This one was a bit more difficult to handle than the others, but thanks to the (thin) simd abstraction, everything went well.
> 
> The patch looks great! quick question w.r.t the SIMD abstraction: does it in anyway overlap with abstracting universal intrinsics using C++ templates? I am not sure what the scope of that work is, so it might not be a relevant question at all :)

At some point I'd love to use the xsimd abstraction instead, but that's more of a midterm goal, so the slight abstraction here is fine (to me).

serge-sans-paille commented:
It's just a translation from C.src to C++, I'm not adding (nor removing !) architecture support etc, in the spirit of the previous PR on the *road to C++*. In a similar manner, I'm not changing the original algorithm and don't plan to (well, that's a great topic and I'd love investing time on it, but that's another story, and not this PR's).

Concerning the genericity, using a generic simd abstraction (like the one provided by xsimd) indeed looks beneficial ;-)

charris commented:
Thanks @serge-sans-paille.

Yes, the questions involve the original. I'd like to see more on that somewhere along the line. There is also the problem of indirect sorts.

r-devulap commented:
@charris The speed up comes really from three sources: 

1) Vectorizing the quicksort partitioning algorithm. AVX-512 _compressstore_ instruction comes in super handy for partitioning and provides a very nice speed up. 
2) Bitonic sorting networks for smaller arrays (<= 128 elements). It parallelizes a lot of the compare and swap. Additionally, the fact that you perform the compare swap operations in the AVX-512 registers means you read and write to memory a lot fewer times. 
3) Picking the median of randomly chosen 72 elements. This isn't as beneficial as the previous two, but still adds value specially when the distribution is uniformly random. 

Hope this helps.

charris commented:
> AVX-512 compressstore instruction comes in super handy for partitioning

I'd like to understand how that works.

Can any of this be generalized for 1) indirect sorts, 2) merging, as in plain old mergesort. I'm more interested in the possibility of having some useful general routines. What would also interesting is some curves of `time/(n*log(n))` vs `log(n)` in an attempt to tease out the value of the different parts. Maybe even some profiling. Is any of that in the original papers? 

What bothers me about the current code is its complexity and restriction to 32 bits on avx512, and perhaps the lack of indirect sorts.
